Using predefined language for embedded explorations

mit0110 opened the pull request:
Hi!
I'm not sure who is going to review this @BenHenning or @seanlip (I know you are OOO for a while), but it is not urgent so it can wait.

I have been working in the i18n for the embedded explorations and finally found a solution, although is not so simple. The main problem is that when angular is using the cookie storage, it basically ignores all your desires if you don't manually set the language clicking in the language footer. The main changes are:

-- There is a global variable GLOBALS.IS_EMBEDDED set in a .run block. There was no way to set it during the .config block, and neither in the i18n controller because it was never called in embedded explorations. If you have a more elegant solution than using global variables, please let me know.
-- The cookie storage is replaced for a custom storage that basically does the same but ignores cookies when the exploration is embedded.
-- Instead of automatically using the exploration language as Sean suggested, the exploration author must add an attribute to the <oppia></oppia> element called site-language. I thought this was a simpler solution for the first try, let me know if the original option is worth implementing.
-- Although the version 0.0.0 of the embedding script is not tested automatically, the changes work for both versions.
-- New test.

Thank you very much!

seanlip commented:
Hi @mit0110, I'll be reviewing this sometime this week. Sorry for the delay!

seanlip commented on static/scripts/oppia-player-0.0.0.js:
```diff
@@ -208,6 +210,9 @@
     }
 
     var iframe = document.createElement('iframe');
+    iframe.isEmbedded = true;
```
Actually, you can't change this script. It's already serving off of CDNs, so if we want to update it we'll have to make a new version. This is a bit annoying to do, so ideally we would have a solution that doesn't involve modifying anything in static/scripts at all.

To check whether the site is embedded, I recommend using the URL. See UrlService in app.js -- it has a method called isIframed().

mit0110 replied:
Now it makes much more sense that all this files are under a "static" folder... sorry for that. I'm reverting the changes!

seanlip commented on static/scripts/oppia-player-0.0.1.js:
```diff
@@ -165,7 +165,8 @@
       var fixedHeight = 'false';
       var fixedWidth = 'false';
-      var explorationVersion = this.oppiaNode.getAttribute('exploration-version') || '';
+      var explorationVersion = this.oppiaNode.getAttribute(
```
Ditto -- we should not make any changes to this file.

seanlip commented on static/scripts/oppia-player-0.0.0.js:
```diff
@@ -208,6 +210,9 @@
 
     var iframe = document.createElement('iframe');
+    iframe.isEmbedded = true;
+    iframe.preferredSiteLanguage = oppiaNode.getAttribute(
```
Not keen on this; I still think the language of the exploration should be used.

mit0110 replied:
Done! It was more simple than I though with the isIframed function.

seanlip commented on static/scripts/embedding_tests_dev_0.0.0.html:
```diff
@@ -2,7 +2,7 @@
 
 <h3>Standard embedding of the latest version</h3>
-<oppia oppia-id="0"></oppia>
+<oppia oppia-id="0" site-language="es"></oppia>
```
I wouldn't add site-language here.

seanlip commented on static/scripts/embedding_tests_dev_0.0.1.html:
```diff
@@ -16,7 +16,7 @@
 <div class="protractor-test-deferred">
   <h3>Standard embedding of version 1 of the exploration with deferred loading</h3>
-  <oppia oppia-id="12" autoload="false" exploration-version="1"></oppia>
+  <oppia oppia-id="12" autoload="false" exploration-version="1" site-language="es"></oppia>
```
Ditto.

seanlip commented on core/tests/protractor/embedding.js:
```diff
@@ -142,9 +142,46 @@ describe('Embedding', function() {
+    var driver = browser.driver;
+    driver.get(
+      general.SERVER_URL_PREFIX + general.SCRIPTS_URL_SLICE +
+      'embedding_tests_dev_0.0.1.html');
```
If you like, you could also make a separate test page for the i18n tests in particular?

mit0110 replied:
How would you name that file? Because we already have i18n.js and embedding.js. embeddingI18n.js wouldn't be redundant? Do you think is better to move the test to the i18n.js file?

seanlip replied:
Oh, sorry -- I meant replacing embedding_tests_dev_0.0.1.html with another HTML file dedicated to testing i18n. The e2e test can still stay here.

mit0110 replied:
Oh that makes more sense. This one works perfectly for now!

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -60,6 +61,26 @@ oppia.controller('I18nFooter', [
+    },
+    get: function(name) {
+      // If the variable is not set assume true
+      if (GLOBALS.IS_EMBEDDED || GLOBALS.IS_EMBEDDED === undefined) {
```
Consider referring to urlService.isIframed() instead. We try to treat GLOBALS as a constant.

mit0110 replied:
Yes I like that option much more! I completely missed this isIframed function, thanks for pointing it out!

seanlip commented on core/templates/dev/head/i18n.js:
```diff
@@ -83,13 +104,24 @@ oppia.config([
     .forceAsyncReload(true);
-}]);
+}]).run(function($translate, $window) {
+  // Set the site language in case the exploration is embedded in an iframe.
```
It's possible that some of this might not be needed, if we use Oppia's built-in iframe detection.

mit0110 replied:
Deleting...

seanlip commented:
Hi @mit0110, I'm really sorry for the delay in replying!

I think the custom cookie storage idea is nice and elegant. However, I lean against using GLOBALS.IS_EMBEDDED (it seems a bit fragile) and I also would still recommend using the exploration language. At which point does this need to be supplied, though? If it needs to be supplied at the outset, we could perhaps surface it from the backend controller in core.controllers.reader.ExplorationPage (note that this is the controller that handles the initial page load) by passing it along via GLOBALS. I'd also prefer not to modify any of the embedding scripts if possible (more details in the individual commits).

Please let me know if you have any questions about this. Thanks!

seanlip commented on core/tests/protractor/embedding.js:
```diff
@@ -142,9 +142,46 @@ describe('Embedding', function() {
 
-  afterEach(function() {
-    general.checkForConsoleErrors([]);
+  it('should have the site language or English as default', function() {
```
Nice test :-)

mit0110 commented:
Hi Sean!

I've basically changed everything again, sorry for the mess with the static files. I learned the lesson! In the last commit, the exploration language is set as site language for embedded frames and the GLOBALS variable is not used anymore. This solved a lot of problems, and now we don't need a custom cookie storage.

Thanks for the review!
Mili

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -441,6 +441,13 @@ oppia.directive('conversationSkin', [function() {
             $rootScope.loadingMessage = '';
             $scope.hasFullyLoaded = true;
 
+            // If the exploration is embedded, use as site language the
```
Suggest: "... use the exploration language as the site language."

mit0110 replied:
Done!

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -441,6 +441,13 @@ oppia.directive('conversationSkin', [function() {
+            // exploration language.
+            var siteLanguage = oppiaPlayerService.getExplorationLanguageCode();
+            if ($window.GLOBALS.SUPPORTED_SITE_LANGUAGES && $scope.isIframed) {
+              $translate.use(oppiaPlayerService.getExplorationLanguageCode());
```
Let's say the exploration language code is not in GLOBALS.SUPPORTED_SITE_LANGUAGES. What happens?

mit0110 replied:
English is used as default. I added the call explicitly (just in case) and some more comments.
I'm just stating what the code will do, do we want this behavior, right?

seanlip replied:
Yep, this behavior sounds fine. Thanks for clarifying!

seanlip commented:
Just a couple of comments, otherwise looks good to merge. Thanks @mit0110!

mit0110 commented:
Hi Sean! In the last commit I addressed your comments and also fixed the end to end tests that were broken. (I forgot to run them all together before the previous commit)

Thanks!
Mili

seanlip commented on core/tests/protractor/embedding.js:
```diff
@@ -146,47 +146,40 @@ describe('Embedding', function() {
     admin.reloadExploration('protractor_test_1.yaml');
 
-    // Change language to Spanish
+    // Change language to Thai, wich is not supported as site language
```
nit: which

seanlip commented on core/tests/protractor/i18n.js:
```diff
@@ -47,50 +48,51 @@ describe('Site language', function() {
-    element.all(by.css('.select2-drop-active li div')).each(function(element) {
-      element.getText().then(function(text) {
+    element(by.css('.protractor-test-system-language-selector')).click();
+    element.all(by.css('.select2-drop-active li div')).each(function(node) {
```
I think we generally use something like 'optionElem; instead of 'node'

mit0110 replied:
Done!

seanlip commented on static/scripts/README:
```diff
@@ -5,10 +5,12 @@ This folder contains embedding scripts for Oppia explorations, and tests for
+- Load the 'Welcome' demo exploration (with id '0') for version 0.0.0 and
+  the 'protractor_test_1' exploration (with id '12') for version 0.0.1
+  on a local development server.
+- Change the content of the initial state to anything you like. Change also the
```
This is just for the protractor_test_1 exploration, right?

Also, perhaps this couple of paragraphs is a bit obsolete, since we now have the e2e tests. Maybe just delete all the stuff from "TESTING INSTRUCTIONS" all the way to "embedding_tests_dev_0.0.1.html"?

mit0110 replied:
I found this particularly helpful when I was testing and to understand how explorations are embedded in other sites. I would leave it as an example, but I changed a the comment to reflect the fact that there are automatic tests now. Also, I removed the TODO. Let me know it if need further changes.

seanlip replied:
Ah ok. What you're suggesting sounds fine to me!

seanlip commented on core/templates/dev/head/player/ConversationSkinDirective.js:
```diff
@@ -441,11 +441,15 @@ oppia.directive('conversationSkin', [function() {
+            // as site language, English is used as default.
             var siteLanguage = oppiaPlayerService.getExplorationLanguageCode();
-            if ($window.GLOBALS.SUPPORTED_SITE_LANGUAGES && $scope.isIframed) {
+            if ($window.GLOBALS.SUPPORTED_SITE_LANGUAGES[siteLanguage] &&
```
Just to double check: previously, you checked for the existence of $window.GLOBALS.SUPPORTED_SITE_LANGUAGES. Was that an error (i.e. is GLOBALS.SUPPORTED_SITE_LANGUAGES always present)? If not, we may still need to check for it here, in order  to prevent a "cannot read siteLanguage of undefined" console error.

mit0110 replied:
First, it was an error, what I actually wanted to check is $window.GLOBALS.SUPPORTED_SITE_LANGUAGES[siteLanguage], to see if the language selected is supported as site language.
Second, $window.GLOBALS.SUPPORTED_SITE_LANGUAGES is set in the base.html file. All embedded explorations inherit from there, so it should always be present. Do you think we should check anyway?

seanlip replied:
Probably not -- I think this is fine, based on your explanation. Thanks @mit0110!

seanlip commented:
Four more comments based on your previous commit (sorry)!

seanlip commented:
LGTM, thanks @mit0110! Merging.
